### ISSUE TITLE
Early stopping mechanism

### ISSUE THREAD

**Issue by 4rtemi5** [tflearn/tflearn#361]
So I hacked together an EarlyStoppingCallback class that does everything it needs to do, but once it is supposed to stop training I don't know how to do that. Could you kindly help me with the last line?

```
class EarlyStoppingCallback(tflearn.callbacks.Callback):
    
    def __init__(self, early_stopping_rounds=None, early_stopping_metric=None, early_stopping_metric_minimize=None):
        self.early_stopping_rounds = early_stopping_rounds
        self.early_stopping_metric = early_stopping_metric
        self.early_stopping_metric_minimize = early_stopping_metric_minimize
        self._best_value_step = None
        self._best_value = None
        self._early_stopped = False

    def on_epoch_end(self, training_state):
        metrics = {
            'accuracy': training_state.global_acc,
            'loss': training_state.global_loss,
            'loss': training_state.loss_value,
            'acc': training_state.acc_value,
            'val_loss': training_state.val_loss,
            'val_acc': training_state.val_acc,
            'epoch': training_state.epoch,
            'step': training_state.current_iter
        }
        #print(metrics)
        #time.sleep(10)
        
        # Early stopping logic.
        if self.early_stopping_rounds is not None:
            if self.early_stopping_metric not in metrics.keys():
                raise ValueError("Metric %s missing from outputs %s." % (
                        self.early_stopping_metric, set(metrics.keys())))
            current_value = metrics[self.early_stopping_metric]
             #print(self._best_value)
             #print(current_value)
            if (self._best_value is None or (self.early_stopping_metric_minimize and (current_value < self._best_value)) or
                (not self.early_stopping_metric_minimize and (current_value > self._best_value))):
                self._best_value = current_value
                self._best_value_step = metrics['epoch']
                print('New best Value: ' + str(self._best_value) + '      at step: ' + str(self._best_value_step))
            stop_now = (metrics['epoch'] - self._best_value_step >= self.early_stopping_rounds)
            if stop_now:
                self._early_stopped = True
                print('Should stop now!')
                time.sleep(10)
                #HOW TO STOP TRAINING HERE??
```
Thanks a lot!

**Comment by aymericdamien** [comment]
There is no mechanism for early stopping yet. An easy way to stop the training would be to raise an Exception (`raise Exception`) when you want your training to stop.
You can have a look at the training loop here: https://github.com/tflearn/tflearn/blob/master/tflearn/helpers/trainer.py#L281

**Comment by shark8me** [comment]
The `on_epoch_end` callback is ideally suited for throwing the exception. It takes the `TrainingState` argument, which has the training/validation loss (or other metrics) available.
